FP16 Support - FP16 scalar value manipulation

SadiinsoSnowfall opened the pull request:
Implement base support for FP16 in EVE.

* `std::float16_t` is only provided in C++23, and currently not implemented by clang. We cannot rely on that.
* The `_Float16` type, and associated softfloat implementations are provided by the compilers on most platforms. We still need to implement some kind of polyfills when it is not provided.

This PR adds the following:
*  A new `eve::float16` type, wrapping a `_Float16` when available with polyfill supports when the compiler does not provide this type. 
* Test changes to allow for adding fp16 to unit tests.
* FP16 constants support.

SadiinsoSnowfall commented:
Quick clarification for the float16 tests: `valmax(f16) + 1f16 == valmax(f16)`, this is because the left side of this comparison, `65505.0f16` is not representable and gets rounded down to `65504.0f`. This is the [expected behaviour](https://godbolt.org/z/a5cz4KcrT).

DenisYaroshevskiy commented on include/eve/arch/float16.hpp:
```diff
@@ -0,0 +1,379 @@
+    }
+
+    template <std::floating_point T>
+    constexpr std::uint16_t emulated_fp_to_fp16(T value) noexcept
```
why are you emulating float 16 by hand

SadiinsoSnowfall commented:
I have changed the representation of `eve::float16` as last discussed. One small problem I see now is that `std::is_trivially_constructible_v<eve::float16, float>` returns different values whether `_Float16` exists or not.

Another problem is that `eve::float16 { 1.0f }` will now trigger the `narrowing` warning on some compiler (I'm not sure why not *all* compilers) when `_Float16` is not available: `ISO C++ does not allow converting to 'eve::float16' {aka '_Float16'} from 'float' with greater conversion rank`

I changed the implementation of the emulation routine to those proposed by Joel, found [here](https://github.com/QiJune/majel/blob/master/float16.h). This implementation does pass the unit tests with a single change (to better handle the expected rounding behaviour towards the max values). A more exhaustive test on all f32 bit patterns reveals that this implementation is quite poor precision-wise, failing to replicate the hardware results on 8% of the bit patterns (mainly on small values and subnormals). I found [another implementation](https://github.com/Maratyszcza/FP16) which is around as performant and succeed in replicating the hw results in 100% of the tests on my machine, but isn't compatible license-wise with EVE (under the MIT license).

DenisYaroshevskiy commented on include/eve/arch/float16.hpp:
```diff
@@ -0,0 +1,283 @@
+{
+  namespace detail
+  {
+    // taken from https://github.com/QiJune/majel/blob/master/float16.h
```
tagging @QiJune just in case.

SadiinsoSnowfall commented on include/eve/arch/float16.hpp:
```diff
@@ -0,0 +1,283 @@
+        constexpr EVE_FORCEINLINE operator long()               const noexcept { return into<long>(); }
+        constexpr EVE_FORCEINLINE operator unsigned long()      const noexcept { return into<unsigned long>(); }
+        constexpr EVE_FORCEINLINE operator long long()          const noexcept { return into<long long>(); }
+        constexpr EVE_FORCEINLINE operator unsigned long long() const noexcept { return into<unsigned long long>(); }
```
Implicit is required for `static_cast` to the target type to work. The native type (`_Float16`) seems to be implicitly convertible to every other arithmetic type. Although this leads to some warning on GCC: https://godbolt.org/z/YobEhsn9c

DenisYaroshevskiy replied:
I don't see the implicit being required https://godbolt.org/z/rYxxcYsa5

SadiinsoSnowfall commented on include/eve/arch/float16.hpp:
```diff
@@ -0,0 +1,283 @@
+    constexpr static std::uint16_t float16_to_bits(float16 value) noexcept
+    {
+      return std::bit_cast<std::uint16_t>(value);
+    }
```
That was there to avoid using `bit_cast` everywhere, but now that the same implementation is used for both the native and emulated type I guess we can just call bit_cast to convert from / to `uint16_t`

SadiinsoSnowfall commented on include/eve/arch/float16.hpp:
```diff
@@ -0,0 +1,283 @@
+    {
+      static constexpr bool supports_f16_type = spy::supports::fp16::type;
+      static constexpr bool supports_f16_scalar_ops = spy::supports::fp16::scalar_ops;
+      static constexpr bool supports_f16_vector_ops = spy::supports::fp16::vector_ops;
```
fixed

SadiinsoSnowfall commented on include/eve/arch/float16.hpp:
```diff
@@ -0,0 +1,283 @@
+      static constexpr bool supports_f16_vector_ops = spy::supports::fp16::vector_ops;
+    }
+
+    using float16 = _Float16;
```
Closing as per our last slack discussion.
